src/acl.c

```diff
@@ -1358,41 +1579,65 @@ int ACLCheckPubsubChannelPerm(sds channel, list *allowed, int literal) {
     return ACL_OK;
 }
 
+int ACLCheckUserPubsubChannelPerm(user *u, sds channel, int literal) {
+    return ACLCheckPubsubChannelPerm(channel, ACLUserGetDefaultSelector(u)->channels, literal);
+}
+
 /* Check if the user's existing pub/sub clients violate the ACL pub/sub
  * permissions specified via the upcoming argument, and kill them if so. */
-void ACLKillPubsubClientsIfNeeded(user *u, list *upcoming) {
+void ACLKillPubsubClientsIfNeeded(user *new, user *original) {
     listIter li, lpi;
     listNode *ln, *lpn;
     robj *o;
     int kill = 0;
     
-    /* Nothing to kill when the upcoming are a literal super set of the original
-     * permissions. */
-    listRewind(u->channels,&li);
-    while (!kill && ((ln = listNext(&li)) != NULL)) {
-        sds pattern = listNodeValue(ln);
-        kill = (ACLCheckPubsubChannelPerm(pattern,upcoming,1) ==
-                ACL_DENIED_CHANNEL);
+    /* Nothing to kill when one of the selectors has all channel permissions. */
+    listRewind(new->selectors,&li);
+    while((ln = listNext(&li))) {
+        aclSelector *s = (aclSelector *) listNodeValue(ln);
+        if (s->flags & SELECTOR_FLAG_ALLCHANNELS) return;
+    }
+
+    list *upcoming;
+    if (listLength(new->selectors) == 1) {
+        upcoming = ACLUserGetDefaultSelector(new)->channels;
+    } else {
+        upcoming = listCreate();
+        listRewind(new->selectors,&li);
+        while((ln = listNext(&li))) {
+            aclSelector *s = (aclSelector *) listNodeValue(ln);
+            listRewind(s->channels, &lpi);
+            while((lpn = listNext(&lpi))) {
+                listAddNodeTail(upcoming, listNodeValue(ln));
+            }
+        }
```

Comment:
please add a comment above the block to describe what it does (compose the `upcoming` list).
.. easier to read and skip when big chunks of code have a title.

Reply:
The logic was rewritten to hopefully be clearer now.